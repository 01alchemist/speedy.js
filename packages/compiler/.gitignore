src/**/*.js
test/**/*.js
cases/**/*.js
*.map
cases/fib.wasm
tools
index.js

