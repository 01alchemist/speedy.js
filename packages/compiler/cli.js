#!/usr/bin/env node
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var path = require("path");
var ts = require("typescript");
var program = require("commander");
var packageJson = require("./package.json");
var diagnostics_1 = require("./src/util/diagnostics");
var compiler_1 = require("./src/compiler");
var speedyjs_compiler_options_1 = require("./src/speedyjs-compiler-options");
function parseCommandLine() {
    function parseSettings(setting, memo) {
        var _a = setting.split("="), key = _a[0], value = _a[1];
        memo[key] = parseInt(value);
        return memo;
    }
    program
        .version(packageJson.version)
        .usage("[options] [files ...]")
        .option("-c --config <configFile>", "The path to the tsconfig.json")
        .option("--unsafe", "Use the unsafe runtime system")
        .option("--emit-llvm", "Emit LLVM Assembly Code instead of WASM files")
        .option("--binaryen-opt", "Optimize using Binaryen opt")
        .option("-s --settings [value]", "additional settings", parseSettings, {})
        .parse(process.argv);
    return program;
}
function parseConfigFile(configFileName) {
    var configurationFileText = ts.sys.readFile(configFileName);
    var jsonConfig = ts.parseConfigFileTextToJson(configFileName, configurationFileText);
    if (jsonConfig.error) {
        diagnostics_1.reportDiagnostics([jsonConfig.error]);
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }
    var parsedConfiguration = ts.parseJsonConfigFileContent(jsonConfig.config, ts.sys, path.dirname(configFileName), undefined, configFileName);
    if (parsedConfiguration.errors.length > 0) {
        diagnostics_1.reportDiagnostics(parsedConfiguration.errors);
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }
    return parsedConfiguration;
}
function getCompilerOptions(commandLine, tsConfigFileName) {
    var rootFileNames = [];
    var compilerOptions;
    if (tsConfigFileName) {
        var configuration = parseConfigFile(tsConfigFileName);
        rootFileNames = configuration.fileNames;
        compilerOptions = configuration.options;
    }
    else {
        rootFileNames = commandLine.files;
        compilerOptions = ts.getDefaultCompilerOptions();
    }
    compilerOptions.unsafe = commandLine.unsafe;
    compilerOptions.binaryenOpt = commandLine.binaryenOpt;
    compilerOptions.emitLLVM = commandLine.emitLlvm;
    compilerOptions.globalBase = commandLine.settings.GLOBAL_BASE;
    compilerOptions.totalMemory = commandLine.settings.TOTAL_MEMORY;
    compilerOptions.totalStack = commandLine.settings.TOTAL_STACK;
    return { rootFileNames: rootFileNames, compilerOptions: speedyjs_compiler_options_1.initializeCompilerOptions(compilerOptions) };
}
function run() {
    var commandLine = parseCommandLine();
    var configFileName = commandLine.config || ts.findConfigFile(ts.sys.getCurrentDirectory(), ts.sys.fileExists);
    if (commandLine.args.length === 0 && !configFileName) {
        ts.sys.write("Pass either a config file or the files to use");
        commandLine.outputHelp();
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }
    var _a = getCompilerOptions(commandLine, configFileName), compilerOptions = _a.compilerOptions, rootFileNames = _a.rootFileNames;
    var compilerHost = ts.createCompilerHost(compilerOptions);
    var compiler = new compiler_1.Compiler(compilerOptions, compilerHost);
    var _b = compiler.compile(rootFileNames), diagnostics = _b.diagnostics, exitStatus = _b.exitStatus;
    if (diagnostics.length > 0) {
        diagnostics_1.reportDiagnostics(diagnostics);
    }
    return ts.sys.exit(exitStatus);
}
run();
//# sourceMappingURL=cli.js.map