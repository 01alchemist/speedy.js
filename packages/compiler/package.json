{
  "name": "speedyjs-compiler",
  "version": "0.0.1",
  "main": "index.js",
  "bin": "index.js",
  "author": "Micha Reiser <micha@famreiser.ch>",
  "license": "MIT",
  "repository": "https://github.com/MichaReiser/speedy.js.git",
  "devDependencies": {
    "@types/jest": "^18.1.1",
    "gulp": "^3.9.1",
    "jest": "^19.0.2",
    "merge-stream": "^1.0.1",
    "ts-jest": "^19.0.1"
  },
  "dependencies": {
    "@types/debug": "^0.0.29",
    "@types/node": "^7.0.5",
    "@types/tmp": "0.0.32",
    "debug": "^2.6.3",
    "llvm-node": "^0.1.1",
    "speedyjs-runtime": "0.x.x",
    "tmp": "0.0.31",
    "typescript": "MichaReiser/TypeScript#release"
  },
  "scripts": {
    "install": "node ./scripts/install-dependencies.js",
    "test": "jest",
    "test:watch": "jest --watch",
    "deploy": "export DEBUG=* && cd cases && node ../index.js && cp fib.wasm /Users/micha/git/webassembly-benchmark/fib-opt.wasm && cd .."
  },
  "jest": {
    "transform": {
      ".(ts)": "<rootDir>/node_modules/ts-jest/preprocessor.js"
    },
    "testResultsProcessor": "<rootDir>/node_modules/ts-jest/coverageprocessor.js",
    "testRegex": "(/__tests__/.*|\\.(test|spec))\\.(ts|js)$",
    "moduleFileExtensions": [
      "ts",
      "js"
    ]
  }
}
