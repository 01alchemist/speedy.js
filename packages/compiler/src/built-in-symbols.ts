import * as ts from "typescript";

/**
 * Resolves the symbols of built in Types like Array, string...
 */
export class BuiltInSymbols {
    private constructor(private stdLibSymbols: Map<string, ts.Symbol>) {
    }

    static create(program: ts.Program, compilerHost: ts.CompilerHost, compilerOptions: ts.CompilerOptions): BuiltInSymbols {
        const symbols = BuiltInSymbols.loadStdLibVariables(program, compilerHost, compilerOptions);
        return new BuiltInSymbols(symbols);
    }

    get(name: string): ts.Symbol | undefined{
        return this.stdLibSymbols.get(name);
    }

    private static loadStdLibVariables(program: ts.Program, compilerHost: ts.CompilerHost, compilerOptions: ts.CompilerOptions): Map<string, ts.Symbol> {
        const defaultLibFileLocation = compilerHost.getDefaultLibLocation!();
        const symbols = new Map<string, ts.Symbol>();
        const stdLibFiles = program.getSourceFiles().filter(file => file.fileName.startsWith(defaultLibFileLocation));

        for (const lib of stdLibFiles) {
            const stdLibVariables = program.getTypeChecker().getSymbolsInScope(lib, ts.SymbolFlags.Variable);
            for (const type of stdLibVariables) {
                symbols.set(type.name, type);
            }
        }

        // else, no stdlib included, so no array... etc
        return symbols;
    }
}
