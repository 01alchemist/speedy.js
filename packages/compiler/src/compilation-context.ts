import * as ts from "typescript";
import * as llvm from "llvm-node";

export interface CompilationContext {
    readonly compilerHost: ts.CompilerHost;
    readonly program: ts.Program;
    readonly compilerOptions: ts.CompilerOptions;
    readonly llvmContext: llvm.LLVMContext;
}
