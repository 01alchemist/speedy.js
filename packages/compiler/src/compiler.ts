import * as ts from "typescript";
import * as llvm from "llvm-node";
import {DefaultCodeGenerationContextFactory} from "./code-generation/default-code-generation-context-factory";
import {PerFileCodeGenerator} from "./code-generation/per-file-code-generator";
import {NotYetImplementedCodeGenerator} from "./code-generation/not-yet-implemented-code-generator";
import {LogUnknownTransformVisitor} from "./transform/log-unknown-transform-visitor";
import {SpeedyJSTransformVisitor} from "./transform/speedyjs-transform-visitor";
import {createTransformVisitorFactory} from "./transform/transform-visitor";

export class Compiler {
    constructor(private compilerOptions: ts.CompilerOptions, private compilerHost: ts.CompilerHost) {
        this.compilerHost = ts.createCompilerHost(compilerOptions);
    }

    compile(rootFileNames: string[], writeFile?: ts.WriteFileCallback): { exitStatus: ts.ExitStatus, diagnostics: ts.Diagnostic[] } {
        const program: ts.Program = ts.createProgram(rootFileNames, this.compilerOptions, this.compilerHost);
        const diagnostics = [...program.getSyntacticDiagnostics(), ...program.getOptionsDiagnostics(), ...program.getGlobalDiagnostics(), ...program.getSemanticDiagnostics() ];

        if (diagnostics.length > 0) {
            return { diagnostics, exitStatus: ts.ExitStatus.DiagnosticsPresent_OutputsSkipped };
        }

        const context = new llvm.LLVMContext();
        const llvmEmitter = new PerFileCodeGenerator(context, new DefaultCodeGenerationContextFactory(new NotYetImplementedCodeGenerator()));

        const logUnknownVisitor = new LogUnknownTransformVisitor();
        const compilationContext = {
            compilerHost: this.compilerHost,
            compilerOptions: this.compilerOptions,
            llvmContext: context,
            program
        };

        const speedyJSVisitor = new SpeedyJSTransformVisitor(compilationContext, llvmEmitter);

        const emitResult = program.emit(undefined, undefined, undefined, undefined, { before: [
            createTransformVisitorFactory(logUnknownVisitor),
            createTransformVisitorFactory(speedyJSVisitor)
        ]});

        llvmEmitter.write();

        const exitStatus = emitResult.emitSkipped ? ts.ExitStatus.DiagnosticsPresent_OutputsSkipped : ts.ExitStatus.Success;
        return { exitStatus, diagnostics: emitResult.diagnostics };
    }
}
