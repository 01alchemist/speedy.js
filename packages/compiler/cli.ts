#!/usr/bin/env node

import * as ts from "typescript";
import * as program from "commander";

const packageJson = require("./package.json");
import {reportDiagnostics} from "./src/util/diagnostics";
import {Compiler} from "./src/compiler";
import {IExportedCommand} from "commander";
import {initializeCompilerOptions, UninitializedSpeedyJSCompilerOptions} from "./src/speedyjs-compiler-options";

interface CommandLineArguments extends IExportedCommand {
    files: string[],
    config?: string,
    unsafe?: boolean;
    emitLlvm?: boolean;
    binaryenOpt?: boolean;
    settings: {
        TOTAL_MEMORY?: number;
        TOTAL_STACK?: number;
        GLOBAL_BASE?: number;
    };
}

function parseCommandLine(): CommandLineArguments {
    function parseSettings(setting: string, memo: any) {
        const [key, value] = setting.split("=");
        memo[key] = parseInt(value);
        return memo;
    }

    program
        .version(packageJson.version)
        .usage("[options] [files ...]")
        .option("-c --config <configFile>", "The path to the tsconfig.json")
        .option("--unsafe", "Use the unsafe runtime system")
        .option("--emit-llvm", "Emit LLVM Assembly Code instead of WASM files")
        .option("--binaryen-opt", "Optimize using Binaryen opt")
        .option("-s --settings [value]", "additional settings", parseSettings, {})
        .parse(process.argv);

    return program as CommandLineArguments;
}

function parseConfigFile(configFileName: string): ts.ParsedCommandLine {
    const configurationFileText = ts.sys.readFile(configFileName);
    const jsonConfig = ts.parseConfigFileTextToJson(configFileName, configurationFileText);
    if (jsonConfig.error) {
        reportDiagnostics([jsonConfig.error]);
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }

    const parsedConfiguration = ts.parseJsonConfigFileContent(jsonConfig.config, ts.sys, ".", undefined, configFileName);
    if (parsedConfiguration.errors.length > 0) {
        reportDiagnostics(parsedConfiguration.errors);
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }

    return parsedConfiguration;
}

function getCompilerOptions(commandLine: CommandLineArguments, tsConfigFileName: string) {
    let rootFileNames: string[] = [];
    let compilerOptions: UninitializedSpeedyJSCompilerOptions;

    if (tsConfigFileName) {
        const configuration = parseConfigFile(tsConfigFileName);
        rootFileNames = configuration.fileNames;
        compilerOptions = configuration.options;
    } else {
        rootFileNames = commandLine.files;
        compilerOptions = ts.getDefaultCompilerOptions();
    }

    compilerOptions.unsafe = commandLine.unsafe;
    compilerOptions.binaryenOpt = commandLine.binaryenOpt;
    compilerOptions.emitLLVM = commandLine.emitLlvm;
    compilerOptions.globalBase = commandLine.settings.GLOBAL_BASE;
    compilerOptions.totalMemory = commandLine.settings.TOTAL_MEMORY;
    compilerOptions.totalStack = commandLine.settings.TOTAL_STACK;

    return { rootFileNames, compilerOptions: initializeCompilerOptions(compilerOptions) };
}

function run() {
    const commandLine = parseCommandLine();

    const configFileName = commandLine.config || ts.findConfigFile(ts.sys.getCurrentDirectory(), ts.sys.fileExists);

    if (commandLine.args.length === 0 && !configFileName) {
        ts.sys.write("Pass either a config file or the files to use");
        commandLine.outputHelp();
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }


    const { compilerOptions, rootFileNames } = getCompilerOptions(commandLine, configFileName);
    const compilerHost = ts.createCompilerHost(compilerOptions);
    const compiler = new Compiler(compilerOptions, compilerHost);

    const { diagnostics, exitStatus } = compiler.compile(rootFileNames);
    if (diagnostics.length > 0) {
        reportDiagnostics(diagnostics);
    }

    return ts.sys.exit(exitStatus);
}

run();
