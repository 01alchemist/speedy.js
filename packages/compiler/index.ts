#!/usr/bin/env node

import * as llvm from "llvm-node";
import * as ts from "typescript";
const packageJson = require("./package.json");
import {reportDiagnostics} from "./src/util/diagnostics";
import {createTransformVisitorFactory} from "./src/transform/transform-visitor";
import {LogUnknownTransformVisitor} from "./src/transform/log-unknown-transform-visitor";
import {SpeedyJSTransformVisitor} from "./src/transform/speedyjs-transform-visitor";
import {PerFileCodeGenerator} from "./src/code-generation/per-file-code-generator";
import {DefaultCodeGenerationContextFactory} from "./src/code-generation/default-code-generation-context-factory";
import {NotYetImplementedCodeGenerator} from "./src/code-generation/not-yet-implemented-code-generator";
import {Compiler} from "./src/compiler";

llvm.initializeAllTargets();
llvm.initializeAllTargetInfos();
llvm.initializeAllAsmPrinters();
llvm.initializeAllTargetMCs();
llvm.initializeAllAsmParsers();

function parseConfigFile(configFileName: string, commandLine: ts.ParsedCommandLine): ts.ParsedCommandLine {
    const configurationFileText = ts.sys.readFile(configFileName);
    const jsonConfig = ts.parseConfigFileTextToJson(configFileName, configurationFileText);
    if (jsonConfig.error) {
        reportDiagnostics([jsonConfig.error]);
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }

    const parsedConfiguration = ts.parseJsonConfigFileContent(jsonConfig.config, ts.sys, ".", commandLine.options, configFileName);
    if (parsedConfiguration.errors.length > 0) {
        reportDiagnostics(parsedConfiguration.errors);
        ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }

    return parsedConfiguration;
}

function run() {
    const commandLine = ts.parseCommandLine(ts.sys.args);
    if (commandLine.errors.length > 0) {
        reportDiagnostics(commandLine.errors);
        return ts.sys.exit(ts.ExitStatus.DiagnosticsPresent_OutputsSkipped);
    }

    if (commandLine.options.version) {
        ts.sys.write(`speedy.js ${packageJson.version}${ts.sys.newLine}`);
        return ts.sys.exit(ts.ExitStatus.Success);
    }

    const configFileName = ts.findConfigFile(ts.sys.getCurrentDirectory(), ts.sys.fileExists);

    if (commandLine.fileNames.length === 0 && !configFileName) {
        ts.sys.write("TODO");
        ts.sys.exit(ts.ExitStatus.Success);
    }

    let rootFileNames: string[] = [];
    let compilerOptions;

    if (configFileName) {
        const configuration = parseConfigFile(configFileName, commandLine);
        rootFileNames = configuration.fileNames;
        compilerOptions = configuration.options;
    } else {
        rootFileNames = commandLine.fileNames;
        compilerOptions = commandLine.options;
    }

    const compilerHost = ts.createCompilerHost(compilerOptions);
    const compiler = new Compiler(compilerOptions, compilerHost);

    const { diagnostics, exitStatus } = compiler.compile(rootFileNames);
    if (diagnostics.length > 0) {
        reportDiagnostics(diagnostics);
    }

    return ts.sys.exit(exitStatus);
}

run();
