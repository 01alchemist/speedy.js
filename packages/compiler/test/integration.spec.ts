import * as fs from "fs";
import * as path from "path";
import * as ts from "typescript";
import * as llvm from "llvm-node";
import * as tmp from "tmp";
import {Compiler} from "../src/compiler";

const CASE_REGEX = /\.case\.ts$/;
const TMP_FILE_REGEX = /\.case\.(js|ll)$/;
const TEST_CASES_DIR = path.resolve(__dirname, "./cases");

describe("Compiler Integration", () => {

    let tmpDirectory: tmp.SynchrounousResult;

    beforeAll(() => {
        llvm.initializeAllTargets();
        llvm.initializeAllTargetInfos();
        llvm.initializeAllAsmPrinters();
        llvm.initializeAllTargetMCs();
        llvm.initializeAllAsmParsers();
    });

    const cases = fs.readdirSync(TEST_CASES_DIR).filter(file => file.match(CASE_REGEX)).map(file => path.join(TEST_CASES_DIR, file));

    for (const testCase of cases) {
        test(path.basename(testCase).replace("case.ts", ""), () => {
            const options = { };
            const compilerHost = ts.createCompilerHost(options);
            const compiler = new Compiler(options, compilerHost);

            const { diagnostics, exitStatus } = compiler.compile([testCase]);

            expect(diagnostics).toEqual([]);
            expect(exitStatus).toBe(ts.ExitStatus.Success);

            const llFileName = testCase.replace(".ts", ".ll");

            if (!fs.existsSync(llFileName)) {
                throw new Error(`No .ll file has been generated. Is the function marked with "use speedyjs"?`);
            }

            const ll = fs.readFileSync(llFileName, "utf-8");

            expect(ll).toMatchSnapshot();
        });
    }

    afterAll(() => {
        const trash = fs.readdirSync(TEST_CASES_DIR).filter(file => file.match(TMP_FILE_REGEX)).map(file => path.join(TEST_CASES_DIR, file));
        trash.forEach(file => fs.unlinkSync(file));
        tmpDirectory.removeCallback();
    });
});
