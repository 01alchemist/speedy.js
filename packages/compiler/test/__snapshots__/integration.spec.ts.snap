// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler Integration array-element-access. 1`] = `
"; ModuleID = 'test/cases/array-element-access.case.ts'
source_filename = \\"test/cases/array-element-access.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define i32 @arrayElementAccessCase() {
entry:
  %0 = alloca [5 x i32]
  %array = alloca i4*
  %returnValue = alloca i32
  %1 = getelementptr inbounds [5 x i32], [5 x i32]* %0, i32 0, i32 0
  store i32 1, i32* %1, align 4
  %2 = getelementptr inbounds [5 x i32], [5 x i32]* %0, i32 0, i32 1
  store i32 2, i32* %2, align 4
  %3 = getelementptr inbounds [5 x i32], [5 x i32]* %0, i32 0, i32 2
  store i32 3, i32* %3, align 4
  %4 = getelementptr inbounds [5 x i32], [5 x i32]* %0, i32 0, i32 3
  store i32 4, i32* %4, align 4
  %5 = getelementptr inbounds [5 x i32], [5 x i32]* %0, i32 0, i32 4
  store i32 5, i32* %5, align 4
  %array1 = getelementptr inbounds [5 x i32], [5 x i32]* %0, i32 0, i32 0
  %6 = call i4* @new_array_i32(i32 5, i32* %array1)
  store i4* %6, i4** %array, align 4
  %array2 = load i4*, i4** %array, align 4
  %array.get = call i32 @array_get_i32(i4* %array2, i32 2)
  store i32 %array.get, i32* %returnValue
  br label %returnBlock

returnBlock:                                      ; preds = %entry
  %7 = load i4*, i4** %array
  call void @delete_array_i32(i4* %7)
  %return = load i32, i32* %returnValue
  ret i32 %return
}

declare i4* @new_array_i32(i32, i32*)

declare i32 @array_get_i32(i4*, i32)

declare void @delete_array_i32(i4*)
"
`;

exports[`Compiler Integration array-literal-expression. 1`] = `
"; ModuleID = 'test/cases/array-literal-expression.case.ts'
source_filename = \\"test/cases/array-literal-expression.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define i4* @constantArrayLiteral() {
entry:
  %0 = alloca [4 x i32]
  %returnValue = alloca i4*
  %1 = getelementptr inbounds [4 x i32], [4 x i32]* %0, i32 0, i32 0
  store i32 1, i32* %1, align 4
  %2 = getelementptr inbounds [4 x i32], [4 x i32]* %0, i32 0, i32 1
  store i32 2, i32* %2, align 4
  %3 = getelementptr inbounds [4 x i32], [4 x i32]* %0, i32 0, i32 2
  store i32 3, i32* %3, align 4
  %4 = getelementptr inbounds [4 x i32], [4 x i32]* %0, i32 0, i32 3
  store i32 4, i32* %4, align 4
  %array = getelementptr inbounds [4 x i32], [4 x i32]* %0, i32 0, i32 0
  %5 = call i4* @new_array_i32(i32 4, i32* %array)
  store i4* %5, i4** %returnValue
  br label %returnBlock

returnBlock:                                      ; preds = %entry
  %return = load i4*, i4** %returnValue
  ret i4* %return
}

declare i4* @new_array_i32(i32, i32*)
"
`;

exports[`Compiler Integration branched-function-returning-void. 1`] = `
"; ModuleID = 'test/cases/branched-function-returning-void.case.ts'
source_filename = \\"test/cases/branched-function-returning-void.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define void @branchedFunctionReturningVoid(i32 %arg1) {
entry:
  %count = alloca i32
  %arg = alloca i32
  store i32 %arg1, i32* %arg
  %arg2 = load i32, i32* %arg, align 4
  %0 = icmp sgt i32 %arg2, 10
  br i1 %0, label %then, label %else

then:                                             ; preds = %entry
  %count3 = load i32, i32* %count, align 4
  store i32 0, i32* %count
  br label %if-successor

else:                                             ; preds = %entry
  %count4 = load i32, i32* %count, align 4
  %arg5 = load i32, i32* %arg, align 4
  store i32 %arg5, i32* %count
  br label %if-successor

if-successor:                                     ; preds = %else, %then
  %count6 = load i32, i32* %count, align 4
  %1 = add i32 %count6, 1
  store i32 %1, i32* %count
  ret void
}
"
`;

exports[`Compiler Integration branched-function-with-return-value. 1`] = `
"; ModuleID = 'test/cases/branched-function-with-return-value.case.ts'
source_filename = \\"test/cases/branched-function-with-return-value.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define i1 @isPrime(i32 %x1) {
entry:
  %x = alloca i32
  store i32 %x1, i32* %x
  %returnValue = alloca i1
  %x2 = load i32, i32* %x, align 4
  %0 = icmp sle i32 %x2, 2
  br i1 %0, label %then, label %else

then:                                             ; preds = %entry
  store i1 true, i1* %returnValue
  br label %if-successor

else:                                             ; preds = %entry
  store i1 false, i1* %returnValue
  br label %if-successor

if-successor:                                     ; preds = %then, %else
  %return = load i1, i1* %returnValue
  ret i1 %return
}
"
`;

exports[`Compiler Integration for-with-return-statement. 1`] = `
"; ModuleID = 'test/cases/for-with-return-statement.case.ts'
source_filename = \\"test/cases/for-with-return-statement.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define i1 @isPrime(i32 %x1) {
entry:
  %i = alloca i32
  %x = alloca i32
  store i32 %x1, i32* %x
  %returnValue = alloca i1
  %x2 = load i32, i32* %x, align 4
  %0 = icmp sle i32 %x2, 2
  br i1 %0, label %then, label %if-successor

then:                                             ; preds = %entry
  store i1 false, i1* %returnValue
  br label %returnBlock

if-successor:                                     ; preds = %entry
  store i32 2, i32* %i, align 4
  br label %for-header

for-header:                                       ; preds = %if-successor8, %if-successor
  %i3 = load i32, i32* %i, align 4
  %x4 = load i32, i32* %x, align 4
  %1 = sitofp i32 %x4 to float
  %2 = call float @llvm.sqrt.f32(float %1)
  %3 = fptosi float %2 to i32
  %4 = icmp slt i32 %i3, %3
  br i1 %4, label %for-body, label %for-successor

for-body:                                         ; preds = %for-header
  %x5 = load i32, i32* %x, align 4
  %i6 = load i32, i32* %i, align 4
  %5 = srem i32 %x5, %i6
  %6 = icmp eq i32 %5, 0
  br i1 %6, label %then7, label %if-successor8

then7:                                            ; preds = %for-body
  store i1 false, i1* %returnValue
  br label %returnBlock

if-successor8:                                    ; preds = %for-body
  %i9 = load i32, i32* %i, align 4
  %7 = add i32 %i9, 1
  store i32 %7, i32* %i
  br label %for-header

for-successor:                                    ; preds = %for-header
  store i1 true, i1* %returnValue
  br label %returnBlock

returnBlock:                                      ; preds = %for-successor, %then7, %then
  %return = load i1, i1* %returnValue
  ret i1 %return
}

; Function Attrs: nounwind readnone
declare float @llvm.sqrt.f32(float) #0

; Function Attrs: nounwind readnone
declare double @llvm.sqrt.f64(double) #0

attributes #0 = { nounwind readnone }
"
`;

exports[`Compiler Integration if-else-if-else. 1`] = `
"; ModuleID = 'test/cases/if-else-if-else.case.ts'
source_filename = \\"test/cases/if-else-if-else.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define i32 @cmp(i32 %x1, i32 %y2) {
entry:
  %result = alloca i32
  %x = alloca i32
  store i32 %x1, i32* %x
  %y = alloca i32
  store i32 %y2, i32* %y
  %returnValue = alloca i32
  %x3 = load i32, i32* %x, align 4
  %y4 = load i32, i32* %y, align 4
  %0 = icmp slt i32 %x3, %y4
  br i1 %0, label %then, label %else

then:                                             ; preds = %entry
  %result5 = load i32, i32* %result, align 4
  store i32 -1, i32* %result
  br label %if-successor

else:                                             ; preds = %entry
  %x6 = load i32, i32* %x, align 4
  %y7 = load i32, i32* %y, align 4
  %1 = icmp eq i32 %x6, %y7
  br i1 %1, label %then8, label %else10

then8:                                            ; preds = %else
  %result9 = load i32, i32* %result, align 4
  store i32 0, i32* %result
  br label %if-successor

else10:                                           ; preds = %else
  %result11 = load i32, i32* %result, align 4
  store i32 1, i32* %result
  br label %if-successor

if-successor:                                     ; preds = %then8, %else10, %then
  %result12 = load i32, i32* %result, align 4
  store i32 %result12, i32* %returnValue
  br label %returnBlock

returnBlock:                                      ; preds = %if-successor
  %return = load i32, i32* %returnValue
  ret i32 %return
}
"
`;

exports[`Compiler Integration if-statement. 1`] = `
"; ModuleID = 'test/cases/if-statement.case.ts'
source_filename = \\"test/cases/if-statement.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define double @max(double %x1, double %y2) {
entry:
  %x = alloca double
  store double %x1, double* %x
  %y = alloca double
  store double %y2, double* %y
  %returnValue = alloca double
  %x3 = load double, double* %x, align 8
  %y4 = load double, double* %y, align 8
  %0 = fcmp ult double %x3, %y4
  br i1 %0, label %then, label %if-successor

then:                                             ; preds = %entry
  %y5 = load double, double* %y, align 8
  store double %y5, double* %returnValue
  br label %returnBlock

if-successor:                                     ; preds = %entry
  %x6 = load double, double* %x, align 8
  store double %x6, double* %returnValue
  br label %returnBlock

returnBlock:                                      ; preds = %if-successor, %then
  %return = load double, double* %returnValue
  ret double %return
}
"
`;

exports[`Compiler Integration initialized-variable-declaration. 1`] = `
"; ModuleID = 'test/cases/initialized-variable-declaration.case.ts'
source_filename = \\"test/cases/initialized-variable-declaration.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define void @initializedVariableDeclaration() {
entry:
  %x = alloca i32
  store i32 10, i32* %x, align 4
  ret void
}
"
`;

exports[`Compiler Integration uninitialized-variable-declaration. 1`] = `
"; ModuleID = 'test/cases/uninitialized-variable-declaration.case.ts'
source_filename = \\"test/cases/uninitialized-variable-declaration.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define void @uninitializedVariableDeclaration() {
entry:
  %x = alloca double
  ret void
}
"
`;

exports[`Compiler Integration void-function-with-return-statement. 1`] = `
"; ModuleID = 'test/cases/void-function-with-return-statement.case.ts'
source_filename = \\"test/cases/void-function-with-return-statement.case.ts\\"
target datalayout = \\"e-m:e-p:32:32-i64:64-n32:64-S128\\"
target triple = \\"wasm32-unknown-unknown\\"

define void @voidFunctionWithReturnStatement(i32 %arg1) {
entry:
  %count = alloca i32
  %arg = alloca i32
  store i32 %arg1, i32* %arg
  %arg2 = load i32, i32* %arg, align 4
  %0 = icmp slt i32 %arg2, 10
  br i1 %0, label %then, label %if-successor

then:                                             ; preds = %entry
  br label %returnBlock

if-successor:                                     ; preds = %entry
  %arg3 = load i32, i32* %arg, align 4
  store i32 %arg3, i32* %count, align 4
  br label %returnBlock

returnBlock:                                      ; preds = %if-successor, %then
  ret void
}
"
`;
